beacon-chain/core/blob/sidecar.go

```diff
@@ -194,29 +194,58 @@ func vectorLinComb(blobs []*v1.Blob, scalars []bls.Fr) ([]bls.Fr, error) {
 	return r, nil
 }
 
-// evaluatePolynomialInEvaluationForm implement the function evaluate_polynomial_in_evaluation_form
-// from the EIP-4844 spec
-func evaluatePolynomialInEvaluationForm(poly []bls.Fr, x *bls.Fr) *big.Int {
-	var tmp, tmpSub, r bls.Fr
+func blsModInv(out *big.Int, x *big.Int) {
+	if len(x.Bits()) != 0 { // if non-zero
+		out.ModInverse(x, &blsModulus)
+	}
+}
+
+// Evaluate a polynomial (in evaluation form) at an arbitrary point `x`
+// Uses the barycentric formula:
+// 	 f(x) = (1 - x**WIDTH) / WIDTH  *  sum_(i=0)^WIDTH  (f(DOMAIN[i]) * DOMAIN[i]) / (x - DOMAIN[i])
+func EvaluatePolyInEvaluationForm(yFr *bls.Fr, poly []bls.Fr, x *bls.Fr) {
+	if len(poly) != params.FieldElementsPerBlob {
+		panic("invalid polynomial length")
+	}
+
+	width := big.NewInt(int64(params.FieldElementsPerBlob))
+	var inverseWidth big.Int
+	blsModInv(&inverseWidth, width)
+
+	// Precomputing the mod inverses as a batch is alot faster
+	invDenom := make([]bls.Fr, params.FieldElementsPerBlob)
+	for i := range invDenom {
+		bls.SubModFr(&invDenom[i], x, &rootsOfUnity[i])
+	}
+	bls.BatchInvModFr(invDenom)
+
+	var y bls.Fr
 	for i := 0; i < params.FieldElementsPerBlob; i++ {
-		bls.SubModFr(&tmpSub, x, &rootsOfUnity[i])
-		bls.MulModFr(&tmp, &poly[i], &rootsOfUnity[i])
-		bls.DivModFr(&tmp, &tmp, &tmpSub)
-		bls.AddModFr(&r, &r, &tmp)
+		var num bls.Fr
+		bls.MulModFr(&num, &poly[i], &rootsOfUnity[i])
+
+		var denom bls.Fr
+		bls.SubModFr(&denom, x, &rootsOfUnity[i])
+
+		var div bls.Fr
+		bls.MulModFr(&div, &num, &invDenom[i])
+
+		var tmp bls.Fr
+		bls.AddModFr(&tmp, &y, &div)
+		bls.CopyFr(&y, &tmp)
 	}
-	// seems PowModInv() isn't implemented in go-kzg except for the pure bignum impl so we have
-	// to convert to bigint here for the final computation. TODO: add this to go-kzg
-	var yChallenge, width, inv, t, xc big.Int
-	width.SetInt64(params.FieldElementsPerBlob)
-	inv.ModInverse(&width, &blsModulus)
-	frToBig(&xc, x)
-	frToBig(&yChallenge, &r)
-	t.Exp(&xc, &width, &blsModulus).
-		Sub(&t, big.NewInt(1)).
-		Mul(&t, &inv)
-	yChallenge.Mul(&yChallenge, &t).
-		Mod(&yChallenge, &blsModulus)
-	return &yChallenge
+
+	xB := new(big.Int)
+	frToBig(xB, x)
+	powB := new(big.Int).Exp(xB, width, &blsModulus)
+	powB.Sub(powB, big.NewInt(1))
+
+	// TODO: add ExpModFr to go-kzg
+	var yB big.Int
+	frToBig(&yB, &y)
+	yB.Mul(&yB, new(big.Int).Mul(powB, &inverseWidth))
+	yB.Mod(&yB, &blsModulus)
+	bls.SetFr(yFr, yB.String())
 }
```

Comment:
This is the same code now as Geth's https://github.com/mdehoog/go-ethereum/blob/eip-4844/crypto/kzg/util.go#L45-L88, with `DomainFr` -> `rootsOfUnity` since that's what Prysm calls it.